### PR TITLE
Update sming 3.9

### PR DESCRIPTION
**Pull request by mikee47**
To try this out, you'll need my `dev/deprecate-delegates` branch from https://github.com/mikee47/Sming.git

### CODE REVIEW

**Comment by mikee47** [comment]
I found a strange with this. Try `make ENABLE_GDB=1` and I get this:

```
r:/sandboxes/esp_rgbww_firmware/out/Esp8266/debug/lib/clib-App-d368dbc221bcd658e6ecbdb532e2b09e.a(webserver.o): In function `ApplicationWebserver::onIndex(HttpRequest&, HttpResponse&)':
r:\sming-v4\sming\core\data/arduinojson.h:86: undefined reference to `std::placeholders::_2'
r:\sming-v4\sming\core\data/arduinojson.h:86: undefined reference to `std::placeholders::_1'
```

This only happens with `ENABLE_GDB=1`, but goes away if I comment-out lines 48-60 in `ApplicationWebserver::init()`. That leaves 7 std::bind calls, any more than that and the error occurs.

I updated everything so it builds for the emulator (`make -j SMING_ARCH=Host`) and no problem, so I'm leaning towards a compiler bug.  Could be a sticking point for std::function?

**Comment by verybadsoldier** [comment]
Thanks, highly appreciated! Looks good so far but will need some more minor tweaks to work I guess (on my side ofc). Could you please rebase this to branch `sming-3.9.x` which I have created for this task?

Some notes:
* Including the RGBWW library by `<RGBWWLed/RGBWWLed.h>` has to be changed to `<Library/RGBWWLed/RGBWWLed.h>`. If that is correct it could be added to the migration guide.
* The placeholder namespace `std::placeholders` is not imported ootb so `_1` is unknown. Maybe the namespace should be imported by default?

**Comment by mikee47** [comment]
Hi, looks like it's already based on `sming-3.9.x` (even with `master`), is that OK?

Firstly, I've removed `Sming` as a submodule - I suspect you did that to accommodate the RGBWWLed library? If that's the only reason, no need. You can clone the Sming repo. separately somewhere and set `SMING_HOME` accordingly as per regs. (Incidentally, do you favour Linux, Windows, or other for dev.? I can tailor my responses accorginly.)

>
>  Including the RGBWW library by `<RGBWWLed/RGBWWLed.h>` has to be changed to `<Library/RGBWWLed/RGBWWLed.h>`. If that is correct it could be added to the migration guide.

Well, it doesn't _require_ changing (I'll explain why shortly) but you do raise a good point, we do need to be consistent about this. I'm thinking we should drop the Libraries/ requirement, but I'll leave that for now.

I've added a new 'Components' directory to the project (to contain private libraries), and added `RGBWWLed` as submodule of the project there. We tell Sming about this in the project's `component.mk` file (which replaces `Makefile-user.mk`). This way, you don't need to muck around with the Sming repo. (You can call it anything, maybe `Libraries` would be better.)

> 
>  The placeholder namespace `std::placeholders` is not imported ootb so `_1` is unknown. Maybe the namespace should be imported by default?

I added these lines to `WiringFrameworkDependencies.h`:
```
#include <functional>
using namespace std::placeholders;
```
Which should cover it. The funny thing is it's not the compiler but the linker that complains...

**Comment by mikee47** [comment]
We're a bit ahead of the curve here but if we can get your project working will be a really good test for Sming. I'll dig out some LED strip this week for a disco :-)

**Comment by verybadsoldier** [comment]
> Hi, looks like it's already based on `sming-3.9.x` (even with `master`), is that OK?

Hm, should not be the case. As far as I can see that branch is based on the `4.2.0-rc1` tag and the included Sming is `3.8.0` with some custom commits? Am I wrong?

> Firstly, I've removed `Sming` as a submodule - I suspect you did that to accommodate the RGBWWLed library? If that's the only reason, no need. You can clone the Sming repo. separately somewhere and set `SMING_HOME` accordingly as per regs. (Incidentally, do you favour Linux, Windows, or other for dev.? I can tailor my responses accorginly.)

Being able to include RGBWWLed without having to integrate it into Sming itself is REALLY good news! But I would still like to keep Sming as a submodule so there is no doubt which exact version of Sming you need and it is a no brainer by just using `git submodule`.

> I've added a new 'Components' directory to the project (to contain private libraries), and added `RGBWWLed` as submodule of the project there. We tell Sming about this in the project's `component.mk` file (which replaces `Makefile-user.mk`). This way, you don't need to muck around with the Sming repo. (You can call it anything, maybe `Libraries` would be better.)

That's really awesome! I have just one problem left regarding MQTT send delays and then I could go for vanilla Sming!

> We're a bit ahead of the curve here but if we can get your project working will be a really good test for Sming. I'll dig out some LED strip this week for a disco :-)

Sounds great but you will also need a bit of hardware components! This is the current version of the hardware project for the firmware I am maintaining:
https://github.com/pljakobs/esp_rgbww_controller/tree/v2.3

Tell me if you need the hardware. I can surely send you one of my spare controllers. Would be version 1.5 since I only got 3 of the new ones myself. But it's barely no difference anyway.

**Comment by mikee47** [comment]
There's updates to all three repos. (`esp_rgbww_firmware`, `RGBWWLed`, `Sming`). Maybe submitting a PR to your repo. isn't the right way forward? I've updated my dev. branch to pull in submodules from my own repo. (and Sming as a submodule a requested), so if you clone that it should just work. This is a snapshot of my dev if that helps:

image.

You can try out the emulator with `make -j SMING_ARCH=Host`, then `make flash run`. Should be helpful for the networking but probably less so for the real-time stuff.

I've had a look at the schematic for the circuit and I can rig up a test jig for that easily enough, so should be fine for hardware. Thanks for the offer, though.

**Comment by verybadsoldier** [comment]
Hm, not sure we are not the same page here regarding the branching stuff: your picture looks ok to me. The `master` in `esp_rgbww_firmware` should currently work with 3.8 only. I still think if you could rebase this PR to branch `sming-3.9.x` then I would merge it and use that branch to work on the port to 3.9 there. Most of the work has been done by you already it seems.

Emulator seems interesting, I will try.

**Comment by mikee47** [comment]
I've renamed the branch to `sming-3.9.x`, is that what you're after?

**Comment by verybadsoldier** [comment]
Sorry if I was unclear: To keep `master` stable and buildable (will probably have to provide a bugfix release soon still based on 3.8) I would prefer this PR to be targeting the `sming-3.9.x` branch (instead of `master`). So I can merge your stuff into that branch and keep working on it without affecting `master`.
Then when that branch is stable enough (against 3.9) then I will merge it into `master`.

Basically like this:
https://help.github.com/en/articles/changing-the-base-branch-of-a-pull-request

(I know you are sound with GitHub. Just to clarify what I mean)

**Comment by mikee47** [comment]
Thank you. I didn't know about that!

**Comment by verybadsoldier** [comment]
Thanks!

**Comment by verybadsoldier** [comment]
@mikee47 
I was just about to get into this sming-3.9 branch and learned that you included Sming using your branch `Sming-rgbww`. I am unsure how to go forward from this state. Are you planning to merge that branch into Sming mainline?
I am tempted to try to build the firmware branch `sming-3.9.x` against the regular Sming `develop` branch but I guess you did the changes in `Sming-rgbww` for some reason.

What would you suggested for me to do here? I would like to start porting the rgbww firmware to new Sming when 3.9 reached a bit of stability but currently there seems still to be much going on. So not sure if it's good now to port rgbww firmware forward?

**Comment by mikee47** [comment]
Just pushed an update through. Whilst porting to ArduinoJson V6 I added a helper function `Json::getValueChanged` which I'm intending to include in a PR for Sming. For the time being I've just added it to webserver.cpp, so easy to take out when Sming's updated.

Most of the changes I made to Delegates for your project are now reverted as they're not necessary. Specifically, using `std::bind` is not such a good idea after all!

Probably the most important changes are a result of the stricter warnings. In particular, missing return values and using uninitialised variables should never have been turned off to start with!

However, the majority of the changes were moving from ArduinoJson V5 to V6. It was definitely worthwhile, but I think that's the biggest hurdle in porting projects so definitely going to bring back support for version 5 as an option.

**Comment by verybadsoldier** [comment]
> Just pushed an update through.

Hm tbh, I cannot see any recent updates in  `mikee47:update/sming-3.9`?

> Most of the changes I made to Delegates for your project are now reverted as they're not necessary. Specifically, using `std::bind` is not such a good idea after all!

Reverted in regular `develop` you mean? So I should switch to that instead of your `dev/Sming-rgbww`? Just tried to build the current state but missing std::placeholders namespace occured and also std::bind didn't compile in `AtClient`.

My idea would be to use this `mikee47:update/sming-3.9` as base for the rgbww firmware and compile against current Sming `develop` as I consider it as "official" state. Is that ok?


> Probably the most important changes are a result of the stricter warnings. In particular, missing return values and using uninitialised variables should never have been turned off to start with!

Yes, thanks for spotting that! I wasn't aware it was configured to ignore those.

> However, the majority of the changes were moving from ArduinoJson V5 to V6. It was definitely worthwhile, but I think that's the biggest hurdle in porting projects so definitely going to bring back support for version 5 as an option.

I am fine with porting over to V6 if that is the way to go.

**Comment by mikee47** [comment]
> 
> 
> > Just pushed an update through.
> 
> Hm tbh, I cannot see any recent updates in `mikee47:update/sming-3.9`?
> 

I've added it as a PR to your `sming-3.9.x` branch, uses the official Sming develop branch (not mine) and your original (but updated) `RGBWWLed`. Sorry for the confusion!
